util/bloom_test.cc

```diff
@@ -601,6 +602,351 @@ TEST_P(FullBloomTest, OptimizeForMemory) {
   }
 }
 
+TEST_P(FullBloomTest, ReserveBloomAndRibbonFilterConstructionMemory) {
+  const BloomFilterPolicy::Mode mode = GetParam();
+  const bool can_reserve_memory =
+      (mode == BloomFilterPolicy::Mode::kFastLocalBloom) ||
+      (mode == BloomFilterPolicy::Mode::kStandard128Ribbon);
+  constexpr std::size_t kSizeDummyEntry = 256 * 1024;
+  constexpr std::size_t kCacheCapacity = 8 * 1024 * 1024;
+  std::size_t memory_reserved = 0;
+  std::size_t num_dummy_entries = 0;
+  std::size_t memory_reserved_during_builder_contruction = 0;
+
+  BlockBasedTableOptions table_options;
+  LRUCacheOptions lo;
+  lo.capacity = kCacheCapacity;
+  lo.num_shard_bits = 0;  // 2^0 shard
+  lo.strict_capacity_limit = true;
+  std::shared_ptr<Cache> cache(NewLRUCache(lo));
+  table_options.block_cache = cache;
+  table_options.filter_policy.reset(
+      new BloomFilterPolicy(FLAGS_bits_per_key, mode));
+  table_options.reserve_bloom_ribbon_filter_construction_memory = true;
+
+  FilterBuildingContext ctx(table_options);
+  std::shared_ptr<CacheReservationManager> cache_res_mgr(
+      new CacheReservationManager(cache));
+  ctx.cache_res_mgr = cache_res_mgr;
+
+  ASSERT_EQ(cache_res_mgr->GetTotalReservedCacheSize(), 0 * kSizeDummyEntry);
+  ASSERT_EQ(cache->GetPinnedUsage(), 0 * kSizeDummyEntry);
+  std::unique_ptr<FilterBitsBuilder> filter_bits_builder(
+      table_options.filter_policy->GetBuilderWithContext(ctx));
+  if (can_reserve_memory) {
+    // Memory reserved now: 1 or 2 sizeof(std::deque<uint64_t> hash_entries_)
+    // during filter_bits_builder construction
+    memory_reserved_during_builder_contruction =
+        (mode == BloomFilterPolicy::Mode::kFastLocalBloom ? 1 : 2) *
+        sizeof(std::deque<uint64_t>);
+    memory_reserved = memory_reserved_during_builder_contruction;
+  } else {
+    memory_reserved = 0;
+  }
+  num_dummy_entries = static_cast<std::size_t>(
+      std::ceil(memory_reserved / static_cast<double>(kSizeDummyEntry)));
+  EXPECT_EQ(cache_res_mgr->GetTotalReservedCacheSize(),
+            num_dummy_entries * kSizeDummyEntry);
+  EXPECT_GE(cache->GetPinnedUsage(), num_dummy_entries * kSizeDummyEntry);
+  EXPECT_LT(cache->GetPinnedUsage(), (num_dummy_entries + 1) * kSizeDummyEntry);
+
+  const std::size_t num_entries = 10000;
+  char key_buffer[sizeof(int)];
+  for (std::size_t i = 0; i < num_entries; ++i) {
+    filter_bits_builder->AddKey(Key(static_cast<int>(i), key_buffer));
```

Comment:
If you have even a rough bound on peak memory usage by Bloom construction, that suffices to check for hash_entries because that is the overwhelming majority of usage for Bloom.